docs/references/ingress-messages.mdx

```diff
@@ -0,0 +1,82 @@
+
+Some operations, such as transfers, must not be executed multiple times. If a call was `rejected` as described above, it is safe to retry the operation. In all other error cases however, retrying can be risky because the call could still successfully execute, potentially leading to severe security vulnerabilities such as double spending.
+
+In general there are cases where it is impossible to determine if a call was successful. For example, suppose
+* The client loses its connection until the request status has been removed from the state tree. (Recall that ICP will remove the request from the system state tree some time after ingress expiry.)
+* There is no application specific way to tell if the call successfully executed. For example, the canister does not provide a way to find out if the call happened.
+
+In this situation, it is indistinguishable if the call was successful, but all information has already been removed from the system state tree, or the call never made it into the state tree in the first place and hasn't been executed. 
+
+### Recommendations and best practices for secure update call handling
+
+The best way to handle errors as discussed above is to implement safe retries using idempotent calls. Idempotency and several patterns to achieve it are discussed in the [best practices on safe retries and idempotency](/docs/current/developer-docs/smart-contracts/best-practices/idempotency).
+
+It may not always be possible to make call endpoints idempotent. For example, an endpoint that is not under the developer’s control may simply not have this property. In this case, we still suggest retrying as a best effort. However, the guarantees in this case are weaker compared to using idempotent calls.
+
+ICP provides a built-in deduplication feature. Namely, if two calls with identical [request id](/docs/current/references/ic-interface-spec#request-ids) are made, the transaction is processed at most once. For two calls to have the same `request_id`, they must have the same values for all fields `request_type`, `sender`, `ingress_expiry`, `nonce`, `canister_id`, `method_name`, and `arg`. This gives idempotency, but only until [ingress_expiry](/docs/current/references/ic-interface-spec#authentication) is reached, because after the message expires, ICP [may remove the request_id from the system state tree](/docs/current/references/ic-interface-spec#state-tree-request-status). Note that the same request will not be accepted later, as it expired. Thus, such requests can be safely retried analogously to how it is done for [idempotent calls](/docs/current/developer-docs/smart-contracts/best-practices/idempotency#idempotent-canister-apis), see the illustration on that page.
```

Comment:
```suggestion
ICP provides a built-in deduplication feature. Namely, if two calls with identical [request id](/docs/current/references/ic-interface-spec#request-ids) are made, the transaction is processed at most once. For two calls to have the same `request_id`, they must have the same values for all fields `request_type`, `sender`, `ingress_expiry`, `nonce`, `canister_id`, `method_name`, and `arg`. This gives idempotency, but only until the [`ingress_expiry`](/docs/current/references/ic-interface-spec#authentication) is reached, because after the message expires, ICP [may remove the `request_id` from the system state tree](/docs/current/references/ic-interface-spec#state-tree-request-status). Note that the same request will not be accepted later, as it expired. Thus, such requests can be safely retried analogously to how it is done for [idempotent calls](/docs/current/developer-docs/smart-contracts/best-practices/idempotency#idempotent-canister-apis), see the illustration on that page.
```

Reply:
Done.